src/hotstuff/protocol.rs

```diff
@@ -5,124 +5,550 @@
+
+            // Check if I need to re-propose an existing block. This may be required in case a chain of consecutive
+            // views of voting for a validator-set-updating block proposed earlier has been interrupted.
+            if let Some(block_hash) = repropose_block(self.view_info.view, block_tree)? {
+                let block = 
+                    block_tree.block(&block_hash)?
+                    .ok_or(BlockTreeError::BlockExpectedButNotFound{block: block_hash})?;
 
-        // 1. Broadcast a NewView message for self.cur_view to leader.
+                let proposal_msg = HotStuffMessage::proposal(self.config.chain_id, self.view_info.view, block);
+                self.sender_handle.broadcast(proposal_msg);
+                return Ok(())
+            }
 
-        // 2. Update cur_view and cur_leader to view and leader, next_leader tp next_leader.
+            // Otherwise, propose a new block or nudge based on highest_qc.
+            let highest_qc = block_tree.highest_qc()?;
+            match highest_qc.phase {
+                // Produce a block proposal.
+                Phase::Generic | Phase::Decide => {
+                    let (parent_block, child_height) = if highest_qc.is_genesis_qc() {
+                        (None, BlockHeight::new(0))
+                    } else {
+                        let parent_height = 
+                            block_tree.block_height(&highest_qc.block)?
+                            .ok_or(BlockTreeError::BlockExpectedButNotFound{block: highest_qc.block})?;
+                        (
+                            Some(highest_qc.block),
+                            parent_height + 1,
+                        )
+                    };
+                    
+                    let produce_block_request = ProduceBlockRequest::new(
+                        self.view_info.view,
+                        parent_block,
+                        block_tree.app_view(parent_block.as_ref())?,
+                    );
+        
+                    let ProduceBlockResponse {
+                        data,
+                        data_hash,
+                        app_state_updates: _,
+                        validator_set_updates: _,
+                    } = app.produce_block(produce_block_request);
+        
+                    let block = Block::new(child_height, highest_qc, data_hash, data);
 
-        // 3. If I am cur_leader then broadcast a nudge or a proposal.
+                    let proposal_msg = HotStuffMessage::proposal(self.config.chain_id, self.view_info.view, block);
 
-        todo!()
+                    self.sender_handle.broadcast(proposal_msg.clone());
+
+                    if let HotStuffMessage::Proposal(proposal) = proposal_msg {
+                        Event::Propose(ProposeEvent{timestamp: SystemTime::now(), proposal}).publish(&self.event_publisher)
+                    }
+
+                },
+                // Produce a nudge.
+                _ => {
```

Comment:
I think our code would be more robust if we don't use a catch-all pattern (`_`) and instead explicitly match. This advice applies generally, not only here.

The problem with the catch-all pattern is that it hides changes to enums. This is a problem because when we modify an enum (e.g., by adding a new variant, or modifying an existing variant), we need to consider whether code that uses the enum needs to be changed as well. If we have a habit of matching exactly, then it's easy to identify the places in the code that needs to be changed, because the compiler will complain about these places. However, if we use catch-all patterns, we need to carefully go through every place an enum is matched to see if the catch-all is still appropriate, introducing the possibility of programmer error.